File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode p1 = l1, p2 = l2;
        ListNode dummy = new ListNode(-1), p = dummy;
        int carry = 0;
        while (p1!= null || p2!= null || carry > 0) {
            int n = p1 != null && p2 != null? p1.val + p2.val: p1 == null && p2 == null? 0 : p1 == null ? p2.val : p1.val;
            if (carry > 0) {
                 n += carry;
                 carry = 0;
            }
            while (n >= 10) {
                carry++;
                n -= 10;
            }
            p.next = new ListNode(n);
            p = p.next;
            if (p1 != null) p1 = p1.next;
            if (p2 != null) p2 = p2.next;
        }
        return dummy.next;
    }
}

File: 0002-add-two-numbers/0002-add-two-numbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode* result = NULL;
    struct ListNode* tail = NULL;
    int carry = 0;
    while (l1 != NULL || l2 != NULL) {
        int sum = carry;
        if (l1 != NULL) {
        sum += l1->val;
        l1 = l1->next;
        }
        if (l2 != NULL) {
        sum += l2->val;
        l2 = l2->next;
        }
        carry = sum / 10;
        struct ListNode* newNode = malloc(sizeof(struct ListNode));
        newNode->val = sum % 10;
        newNode->next = NULL;
        if (result == NULL) {
        result = tail = newNode;
        } else {
        tail->next = newNode;
        tail = newNode;
        }
    }
    if (carry > 0) {
        struct ListNode* newNode = malloc(sizeof(struct ListNode));
        newNode->val = carry;
        newNode->next = NULL;
        tail->next = newNode;
    }
    return result;
}